src/Windows/dll/local_cache.cpp

```diff
@@ -1,33 +1,412 @@
+}
+
+static std::string sha256(const std::string& input)
+{
+    return sha256(input.length(), input.data());
+}
+
+static std::string get_file_name(const std::string& id)
+{
+    return g_cache_dirname + "\\" + sha256(id);
+}
+
+void local_cache_clear() {
+    init();
+
+    WIN32_FIND_DATA data;
+    std::wstring baseDir(g_cache_dirname.begin(), g_cache_dirname.end());
+    std::wstring searchPattern = baseDir + L"\\*";
+
+    HANDLE hFind = FindFirstFile(searchPattern.c_str(), &data);
+
+    if (hFind != INVALID_HANDLE_VALUE)
+    {
+        do {
+            std::wstring fileName(data.cFileName);
+            if (fileName.compare(L".") != 0 && fileName.compare(L"..") != 0)
+            {
+                std::wstring fullFileName = baseDir + L"\\" + fileName;
+
+                throw_if(!DeleteFileW(fullFileName.c_str()),
+                    "Deleting file failed, error code " + GetLastError());
+            }
+        } while (FindNextFile(hFind, &data));
+        FindClose(hFind);
+    }
+
+    return;
+};
+
+HANDLE OpenHandle(LPCWSTR lpFileName, DWORD dwDesiredAccess, DWORD dwShareMode, LPSECURITY_ATTRIBUTES lpSecurityAttributes,
+    DWORD dwCreationDisposition, DWORD dwFlagsAndAttributes, HANDLE hTemplateFile) {
+    
+    HANDLE file;
+    int i = 0;
+    do {
+        file = CreateFile(lpFileName, dwDesiredAccess, dwShareMode, lpSecurityAttributes,
+            dwCreationDisposition, dwFlagsAndAttributes, hTemplateFile);
+        
+        Sleep(SLEEP_RETRY_MS);
+        i++;
+    } while ((file == INVALID_HANDLE_VALUE) &&
+        (GetLastError() == 32) &&
+        (i < MAX_RETRY));
+
+    return file;
+}
+
+
+void local_cache_add(const std::string& id, time_t expiry, size_t data_size, const void* data)
+{
+    throw_if(id.empty(), "The 'id' parameter must not be empty.");
+    throw_if(data_size == 0, "Data cannot be empty.");
+    throw_if(data == nullptr, "Data pointer must not be NULL.");
+
+    init();
+    CacheEntryHeaderV1 header{};
+    header.version = CACHE_V1;
+    header.expiry = expiry;
+
+    std::string filename = get_file_name(id);
+    std::wstring wfilename(filename.begin(), filename.end());
+
+    HANDLE file = OpenHandle(wfilename.c_str(), GENERIC_WRITE, 0, NULL, CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL, NULL);
+    throw_if(file == INVALID_HANDLE_VALUE, "Open handle failed");
+
+    OVERLAPPED overlapvar = { 0 };
+
+    throw_if(!LockFileEx(file, LOCKFILE_EXCLUSIVE_LOCK, 0, MAXDWORD, MAXDWORD, &overlapvar),
+        "Exclusive write lock operation failed");
+
+    DWORD headerwritten;
+    DWORD datawritten;
+
+    throw_if(!WriteFile(file, &header, sizeof(header), &headerwritten, NULL),
+        "Header write to local cache failed");
+
+    throw_if(!WriteFile(file, data, (DWORD)data_size, &datawritten, NULL),
+        "Data write to local cache failed");
+
+    throw_if(!UnlockFileEx(file, 0, MAXDWORD, MAXDWORD, &overlapvar),
+        "Exclusive write unlock operation failed");
+
+    CloseHandle(file);
```

Comment:
You leak the file handle if any of the "throw_if" calls above throws. That's probably not what you want.

Reply:
In general, when using a "throw" pattern, you need to use RAII techniques to manage resource lifetime (std::unique_ptr, auto-handle classes, etc).

Reply:
I added another throw_if that closes the handles before throwing error